Simul packaging

omar-ocampo opened the pull request:
Added packaging folder.
    Simul spec file to build simul 1.16 from external
    repo.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Simul packaing

    Added packaging folder.
    Simul spec file to build simul 1.16 from external
    repo.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

omar-ocampo commented:
How can I trigger a Jenkins build? It did not start automatically.

brianjmurrell commented:
> How can I trigger a Jenkins build? It did not start automatically.

The repo needs to be added correctly for that to happen.  I will update the settings for you...

It seems you did (eventually?) get a build.  It will build on each PR commit "soon".  Your build failed though:

```
[2021-06-14T17:20:15.865Z] /var/tmp/rpm-tmp.Xj9yxB: line 36: module: command not found
```

You are probably missing "BuildRequires: Lmod" for Leap 15.

omar-ocampo commented:
> > How can I trigger a Jenkins build? It did not start automatically.
> 
> The repo needs to be added correctly for that to happen. I will update the settings for you...
> 
> It seems you did (eventually?) get a build. It will build on each PR commit "soon". Your build failed though:
> 
> ```
> [2021-06-14T17:20:15.865Z] /var/tmp/rpm-tmp.Xj9yxB: line 36: module: command not found
> ```
> 
> You are probably missing "BuildRequires: Lmod" for Leap 15.

Cool thanks, will fix

Commit:
Simul packaging

   When using suse OS, add lua-lmod as a build dependency.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Fixed rpmlint issue

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Optmization patch

   Added 01 so inline function is found.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

omar-ocampo commented:
@brianjmurrell , @phender, build passed, RPMs created for centos: https://build.hpdd.intel.com/job/daos-stack/job/simul/job/PR-1/4/artifact/artifacts/centos7/ and leap: https://build.hpdd.intel.com/job/daos-stack/job/simul/job/PR-1/4/artifact/artifacts/leap15/

brianjmurrell commented on Jenkinsfile:
```diff
@@ -0,0 +1,43 @@
+// I.e. for testing library changes
+//@Library(value="pipeline-lib@your_branch") _
+
+packageBuildingPipeline(['distros' : ['centos7', 'leap15']])
```
All test tools should include a build for EL8 now.

```suggestion
packageBuildingPipeline(['distros' : ['centos7', 'centos8', 'leap15']])
```

omar-ocampo replied:
Fixing

brianjmurrell commented on packaging/Jenkinsfile:
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/groovy
```
This file should be removed, it's only used in the packaging repo to test it.

omar-ocampo replied:
Ok, removing

brianjmurrell commented on packaging/Makefile:
```diff
@@ -0,0 +1,18 @@
+NAME    := libfabric
```
Ditto.  FWIW, from a packaging repo, you can run `$path/to/packaging-repo-clone/install` to copy only the necessary files to the packaging repo.

omar-ocampo replied:
Ok, this removed the following files from packaging folder:
Jenkinsfile
Makefile
README.md
install
packaging

brianjmurrell commented on packaging/Makefile_distro_vars.mk:
```diff
@@ -0,0 +1,108 @@
+DOT     := .
```
This file looks like it's out of date with a packaging repo update that was landed yesterday.  Please refresh packaging/ in this repo.

omar-ocampo replied:
Packging repo was updated to latest commit  01542d2611d80311d67e2e6ddd9d731ca4a010f1 and this file was updated

brianjmurrell commented on packaging/README.md:
```diff
@@ -0,0 +1,8 @@
+# Common Packaging files and scripts
```
This file should be removed.

omar-ocampo replied:
Removed

brianjmurrell commented on packaging/install:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
This file should be removed.

This is the script referred to previously.

omar-ocampo replied:
removed

brianjmurrell commented on packaging/packaging:
```diff
@@ -0,0 +1 @@
+.
```
This file should be removed.

omar-ocampo replied:
removed

brianjmurrell commented on Makefile:
```diff
@@ -0,0 +1,9 @@
+
+include packaging/Makefile_packaging.mk
+
+test:
+	$(call install_repos,$(NAME)@$(BRANCH_NAME):$(BUILD_NUMBER))
+	yum -y install $(NAME)
```
This is already provided by https://github.com/daos-stack/simul/blob/0e45b076d23b192daa5630fddc72bfb84751486f/packaging/Makefile_packaging.mk#L378-L381 so should be removed from here.

```suggestion
```

omar-ocampo replied:
Removing duped code

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+%global daos_major 0
+
```
We're no longer doing this.  It can be removed.

```suggestion
```

omar-ocampo replied:
Line removed

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+%global daos_major 0
+
+%global with_mpich 1
+%global with_openmpi3 1
```
You will also need `with openmpi 1` for EL8.  `with_openmpi3` and `with_openmpi` are mutually exclusive where the former is for EL7 (and Leap15) and the latter is for EL8.  Please see any one of the many other tools that build openmpi and mpich variants for how to structure the addition of EL8.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+
+Name:    simul
+Version: 1.16
+Release: 0%{?commit:.git%{shortcommit}}%{?dist}
```
This should start at 1.

```suggestion
Release: 1%{?commit:.git%{shortcommit}}%{?dist}
```

omar-ocampo replied:
Ok

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+
+%if %{with_mpich}
+%package mpich
+Summary: MACSio for MPICH
```
Copy and pasto?

```suggestion
Summary: Simul for MPICH
```

omar-ocampo replied:
Fixed

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+BuildRequires: mpich-devel%{?_isa}
+%endif
+Requires: %{name}%{?_isa} = %{version}-%{release}
+Provides: %{name}-mpich2-daos-%{daos_major} = %{version}-%{release}
```
```suggestion
```

omar-ocampo replied:
Why removing the provides line?

brianjmurrell replied:
Because that was a historic hack that is no longer necessary.  It's been removed from all other test tools and middleware packages.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+BuildRequires: openmpi3-devel%{?_isa}
+%endif
+Requires: %{name}%{?_isa} = %{version}-%{release}
+Provides: %{name}-openmpi3-daos-%{daos_major} = %{version}-%{release}
```
```suggestion
```

omar-ocampo replied:
Why removing the "Provides" line?

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+%endif
+
+%prep
+%setup -q
+%patch1 -p1
```
Should use `%autosetup` here.

```suggestion
%autosetup -p1
```

omar-ocampo replied:
Thats new to me, will try it

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+done
+
+%files
+%license COPYING
+%if %{with_mpich}
```
It's customary to put a space between _sections_:

```suggestion
%license COPYING

%if %{with_mpich}
```

omar-ocampo replied:
Added new line

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+%if %{with_mpich}
+%files mpich
+%{mpi_libdir}/mpich/bin/*
+%endif
+%if %{with_openmpi3}
```
```suggestion
%endif

%if %{with_openmpi3}
```

omar-ocampo replied:
added

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,105 @@
+%endif
+
+%changelog
+* Tue Jun 08 2021 Omar Ocampo <omar.ocampo.coronado@intel.com> - 0.1
```
Please update the date here to be more recent and the version on the end of the line needs to be $version-$release.

omar-ocampo replied:
ok

brianjmurrell replied:
Also please remember not to resolve reviewers' conversations.

brianjmurrell reviewed (changes requested):
Good first pass.

Commit:
Support for RHEL8 and packaing updates

    Removed unused packaging/ files.
    Updated packaing/ files.
    Added support for RHEL8 build.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

omar-ocampo commented:
Got another compilation error on SuSE, seems to be related with openmpi/openmpi3, debugging.

Commit:
Source mpivars.sh

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Source mpivars.sh only on SuSE

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Debugging suse linker error

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
More debug

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Add libfabric-devel dep

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

brianjmurrell commented on simul.spec:
```diff
@@ -46,7 +49,7 @@ and coherence of parallel filesystems.
 Summary: Simul for MPICH
 %if "%{_host_vendor}" == "suse"
-BuildRequires: mpich-devel%{?_isa} lua-lmod
+BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel glibc
```
You should never have to `BR: glibc`.  It's a base component that no Linux system can function without.

```suggestion
BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel
```

brianjmurrell commented on simul.spec:
```diff
@@ -61,7 +64,7 @@ Simul for MPICH
 Summary: Simul for OpenMPI
 %if "%{_host_vendor}" == "suse"
-BuildRequires: openmpi-devel%{?_isa} lua-lmod
+BuildRequires: openmpi-devel%{?_isa} lua-lmod libfabric-devel glibc
```
Ditto WRT glibc.

brianjmurrell commented on simul.spec:
```diff
@@ -99,6 +102,8 @@ for mpi in %{?mpi_list}
     mkdir $mpi
     %module_load $mpi
+    %source_vars $mpi
+    /sbin/ldconfig -p
```
Debug that can be removed?

brianjmurrell commented on simul.spec:
```diff
@@ -99,6 +102,8 @@ for mpi in %{?mpi_list}
 do
     mkdir $mpi
     %module_load $mpi
+    %source_vars $mpi
```
Given the contents of this file:

```
$ cat /usr/lib64/mpi/gcc/mpich/bin/mpivars.sh
case $PATH in
        */usr/lib64/mpi/gcc/mpich/bin*)
        true
        ;;
        *)
        PATH=/usr/lib64/mpi/gcc/mpich/bin:$PATH
esac


case $MANPATH in
        */usr/lib64/mpi/gcc/mpich/share/man*)
        true
        ;;
        *)
        MANPATH=/usr/lib64/mpi/gcc/mpich/share/man:$MANPATH
        export MANPATH
esac


case $LD_LIBRARY_PATH in
        */usr/lib64/mpi/gcc/mpich/lib64*)
        true
        ;;
        *)
        if [ -z "$LD_LIBRARY_PATH" ]; then
          LD_LIBRARY_PATH=/usr/lib64/mpi/gcc/mpich/lib64
        else
          LD_LIBRARY_PATH="/usr/lib64/mpi/gcc/mpich/lib64:$LD_LIBRARY_PATH"
        fi
        export LD_LIBRARY_PATH
esac
```

and the output from sourcing it:

```
+ source /usr/lib64/mpi/gcc/mpich/bin/mpivars.sh
++ case $PATH in
++ true
++ case $MANPATH in
++ true
++ case $LD_LIBRARY_PATH in
++ true
```

it would seem that sourcing this file is a NOOP that can probably be dropped.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%if %{with_mpich}
+%package mpich
+Summary: Simul for MPICH
+%if "%{_host_vendor}" == "suse"
```
Please use the `suse_version` macro for this.

But in fact it's not clear why this is distro dependent.  Why is libfabric-devel only needed on SuSE?

omar-ocampo replied:
Is using "0%{?suse_version} > 0"  ok to check if running suse? Something similar to rhel would be used

Compiling using mpich fails if libfabric-devel is not installed, only happens on suse

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%if "%{_host_vendor}" == "suse"
+BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel glibc
+%endif
+%if "%{_host_vendor}" == "redhat"
```
And the `rhel` macro for this.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%if %{with_openmpi}
+%package openmpi
+Summary: Simul for OpenMPI
+%if "%{_host_vendor}" == "suse"
```
Ditto.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%if "%{_host_vendor}" == "suse"
+BuildRequires: openmpi-devel%{?_isa} lua-lmod libfabric-devel glibc
+%endif
+%if "%{_host_vendor}" == "redhat"
```
Ditto WRT to `rhel` macro.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%if %{with_openmpi3}
+%package openmpi3
+Summary: Simul for OpenMPI 3
+%if "%{_host_vendor}" == "suse"
```
Ditto WRT `suse_version`.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%if "%{_host_vendor}" == "suse"
+BuildRequires: openmpi3-devel%{?_isa} lua-lmod
+%endif
+%if "%{_host_vendor}" == "redhat"
```
Ditto WRT `rhel`.

brianjmurrell commented on simul.spec:
```diff
@@ -0,0 +1,142 @@
+%endif
+
+%changelog
+* Thu Jun 24 2021 Omar Ocampo <omar.ocampo.coronado@intel.com> - $version-$release
```
Put the actual version number and release here.  And may as well update the date also.

```suggestion
* Mon Jun 248 2021 Omar Ocampo <omar.ocampo.coronado@intel.com> - 1.16-1
```

brianjmurrell reviewed (changes requested):
Looking good.  A few more things from your last update that need fixing.

Commit:
Implemented feedback

   Use suse_version and rhel macros.
   Removed calling mpich vars.
   Removed glibc as a dependency.
   Date updated and set vesion of package.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

brianjmurrell commented on simul.spec:
```diff
@@ -48,10 +47,10 @@ and coherence of parallel filesystems.
 Summary: Simul for MPICH
-%if "%{_host_vendor}" == "suse"
-BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel glibc
+%if (0%{?suse_version} > 0)
```
Versions for distros should be as conservative as needed/known and only made more liberal as the need is discovered. 
 Since we don't know (or support even) versions less than 15.x:

```suggestion
%if (0%{?suse_version} >= 1500)
```

brianjmurrell commented on simul.spec:
```diff
@@ -48,10 +47,10 @@ and coherence of parallel filesystems.
-%if "%{_host_vendor}" == "suse"
-BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel glibc
+%if (0%{?suse_version} > 0)
+BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel
```
Don't repeat common packages:

```suggestion
BuildRequires: lua-lmod libfabric-devel
```

brianjmurrell commented on simul.spec:
```diff
@@ -48,10 +47,10 @@ and coherence of parallel filesystems.
+BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel
 %endif
-%if "%{_host_vendor}" == "redhat"
+%if (0%{?rhel} > 0)
```
Per the previous comment, remove this line:
```suggestion
```
and line 55 so that `BR: mpich-devel%{?_isa}` is common for all distros.

brianjmurrell commented on simul.spec:
```diff
@@ -63,10 +62,10 @@ Simul for MPICH
 Summary: Simul for OpenMPI
-%if "%{_host_vendor}" == "suse"
-BuildRequires: openmpi-devel%{?_isa} lua-lmod libfabric-devel glibc
+%if (0%{?suse_version} > 0)
```
Same comments/changes as above for lines 65-70.

brianjmurrell commented on simul.spec:
```diff
@@ -80,10 +79,10 @@ Simul for openmpi
 %package openmpi3
 Summary: Simul for OpenMPI 3
-%if "%{_host_vendor}" == "suse"
+%if (0%{?suse_version} > 0)
```
Same comments/changes as above for lines 82-87.

Commit:
Implemented feedback

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Fixed %endif

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Re-added mpich

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Add debug information, gnu-mpich module not found

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Re-added  back mpich for rhel

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

omar-ocampo commented:
@phender , @mjean308 , can I get some help to review this?

omar-ocampo commented:
@phender , @mjean308 , can I get some help to review this?

shimizukko commented on Jenkinsfile:
```diff
@@ -0,0 +1,43 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+// To use a test branch (i.e. PR) until it lands to master
+// I.e. for testing library changes
+//@Library(value="pipeline-lib@your_branch") _
```
Are these debug lines?

omar-ocampo replied:
The code is inherit from packaging, cant really tell you what is this about.

shimizukko commented on packaging/Dockerfile.centos.7:
```diff
@@ -0,0 +1,34 @@
+#
+# Copyright 2018-2020, Intel Corporation
```
Is there any reason why you use 2020 instead of 2021? Also, why do you use different years in the files added in the packaging directory (2018-2019, 2018-2020, 2019, etc.)?

omar-ocampo replied:
The code is inherit from packaging, cant really tell you what is this about.

shimizukko commented on packaging/Makefile_packaging.mk:
```diff
@@ -0,0 +1,445 @@
+	touch $@
+endif
+
+# see https://stackoverflow.com/questions/2973445/ for why we subst
```
You might want to explain at a high level because the website could change any day.

omar-ocampo replied:
The code is inherit from packaging, cant really tell you what is this about.

shimizukko reviewed (dismissed):
You might want to make the title more descriptive than "Simul packaging". Also, you might want to describe more about this commit in the commit description such as what is Simul, how it's used with DAOS, what you worked on in this commit. I'm not familiar with this code, so it would be better to have context.

mjean308 commented on simul.spec:
```diff
@@ -0,0 +1,139 @@
+%if (0%{?rhel} >= 7)
+BuildRequires: mpich-devel%{?_isa}
+%endif
+Requires: %{name}%{?_isa} = %{version}-%{release}
```
Both centos and sles require mpich-devel.  I think Brian's comment was to remove it from the if else

mjean308 replied:
```suggestion
BuildRequires: mpich-devel%{?_isa}
Requires: %{name}%{?_isa} = %{version}-%{release}
```

mjean308 commented on simul.spec:
```diff
@@ -0,0 +1,139 @@
+BuildRequires: mpich-devel%{?_isa} lua-lmod libfabric-devel
+%endif
+%if (0%{?rhel} >= 7)
+BuildRequires: mpich-devel%{?_isa}
```
```suggestion

```

mjean308 commented on simul.spec:
```diff
@@ -0,0 +1,139 @@
+%package openmpi3
+Summary: Simul for OpenMPI 3
+%if (0%{?suse_version} >= 1500)
+BuildRequires: openmpi3-devel%{?_isa} lua-lmod
```
same as above remove from the if/else and add it after

mjean308 commented on simul.spec:
```diff
@@ -0,0 +1,139 @@
+%endif
+%if (0%{?rhel} >= 7)
+BuildRequires: openmpi3-devel%{?_isa}
+%endif
```
remove

omar-ocampo commented:
> You might want to make the title more descriptive than "Simul packaging". Also, you might want to describe more about this commit in the commit description such as what is Simul, how it's used with DAOS, what you worked on in this commit. I'm not familiar with this code, so it would be better to have context.

Sounds good, let me edit the commit message

Commit:
Improved README.md

    Added section of what is simul
    Added section of the role of simul within DAOS

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Simplified BuildRequires

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Testing mpich build dep

    With mpich4 we should not require to use mpich-devel.

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Mpich devel for rhel 7

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

Commit:
Mpich devel

Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

omar-ocampo commented:
@mjean308 @shimizukko, could you help with a quick review? This one has an extended README.md and non dupped BuildRequires

shimizukko commented on README.md:
```diff
@@ -1,4 +1,21 @@
+    Simul can let us know what POSIX operations are working or not
+    in a POSIX container.
+
+    There should be an automated test for this.
```
What do you mean by there should be a test? Is it going to be implemented in the future version?

omar-ocampo replied:
Yes, in fact is here: https://github.com/daos-stack/daos/pull/5958 but I need the simul package installed on the test nodes before running it

shimizukko replied:
I see. Then you might want to mention that because it sounds like you're still in middle of your work.

omar-ocampo commented:
@mjean308 @shimizukko, could you help with a quick review? This one has an extended README.md and non dupped BuildRequires
